### ISSUE TITLE
lesionlocator_segment.py hangs when iterating over data_iterator

### ISSUE THREAD

**Issue by Muhamamd-Rehman** [MIC-DKFZ/LesionLocator#11]
Hi,
First of all, thank you for making this excellent tool available. I am a PhD student working on PET-guided CT lesion segmentation, and I would like to evaluate LesionLocator on our dataset.

However, I am encountering an issue in the segmentation script. When running `lesionlocator_segment.py` with a single CT `.nii.gz` file and its corresponding lesion mask, the program consistently hangs during iteration over the data_iterator:

```
for preprocessed in data_iterator:    <--- (Here)
     data = preprocessed['data']
     if isinstance(data, str):
          delfile = data
          data = torch.from_numpy(np.load(data))
          os.remove(delfile)
```

I have tried the following:

1. Reduced `-npp` and `-nps` to 1
2. Ran inference using only a **single fold**
3. Converted my instance lesion mask (labels 1, 2, 3) into a single semantic lesion mask (label 1)

Unfortunately, the script still freezes at the same point.
If you have any suggestions or guidance on how to resolve this, I would greatly appreciate it.

**Comment by mrokuss** [comment]
Hi @Muhamamd-Rehman,

thanks a lot for providing all the details (and even dummy data!), that made it much easier to track down what’s happening.

I was able to “reproduce” the situation on my side, although the script does run through for me. Here’s the output I get with your dummy CT:

```
#######################################################################
Please cite the following paper when using LesionLocator:
Rokuss, M., Kirchhoff, Y., Akbal, S., Kovacs, B., Roy, S., Ulrich, C., ... & Maier-Hein, K. (2025).
LesionLocator: Zero-Shot Universal Tumor Segmentation and Tracking in 3D Whole-Body Imaging. CVPR.
#######################################################################

Loading segmentation model.

 === Predicting dummy_ct === 

 Lesion ID 1: 
100%|███████████████████████████████████████████████████████████████████████████| 1/1 [00:03<00:00,  3.24s/it]

 Lesion ID 2: 
100%|███████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.18s/it]

 Lesion ID 3: 
100%|███████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.10s/it]

done with dummy_ct
```

### What’s actually causing the hang

The issue is not in the iterator itself, it’s in the preprocessing.
Unlike **nnInteractive**, which keeps your original voxel spacing, **LesionLocator** *always* resamples CT + labels to **(1.0, 0.8, 0.8)**.

For your spacing **(0.97, 0.97, 5.0)**, this resampling produces a **massive** volume:

* from `(512, 512, 196)`
* to roughly `(~800, ~600, ~600)` voxels

That’s several hundred million voxels per volume.
On my machine the resampling takes noticeable time and spikes RAM usage sharply.

On your system (1 CPU, 8 GB RAM) this very likely pushes the container into memory pressure → causing the apparent *freeze* right at:

```python
for preprocessed in data_iterator:
```

because the iterator begins by resampling the volume.

### A solution: use the memory-efficient PyTorch resampling backend

By default, LesionLocator uses **skimage** resampling, which is accurate but slow and memory-hungry.
There is a much faster and far more memory-efficient **torch** implementation.
We’ve seen slightly worse resampling quality in some corner cases, but for your use case it should be totally fine.

To enable it, modify the `plans.json` inside your checkpoint:

```
LesionLocatorCheckpoint/LesionLocatorSeg/point_optimized/plans.json
```

Replace the resampling settings with:

```json
"resampling_fn_data": "resample_torch_fornnunet",
"resampling_fn_seg": "resample_torch_fornnunet",
"resampling_fn_data_kwargs": {
    "is_seg": false,
    "memefficient_seg_resampling": true,
    "force_separate_z": false
},
"resampling_fn_seg_kwargs": {
    "is_seg": true,
    "memefficient_seg_resampling": true,
    "force_separate_z": false
},
"resampling_fn_probabilities": "resample_torch_fornnunet",
"resampling_fn_probabilities_kwargs": {
    "is_seg": false,
    "memefficient_seg_resampling": true,
    "force_separate_z": false
}
```

You can also simply use this ready-to-use [plans.json](https://github.com/MIC-DKFZ/LesionLocator/blob/main/documentation/assets/plans_torch_resampling.json) i just pushed and replace yours.

If you plan to run **box-prompt mode**, don’t forget to apply the same patch to:

```
LesionLocatorCheckpoint/LesionLocatorSeg/bbox_optimized/plans.json
```

Let me know if switching to the torch resampler resolves the freeze on your Kubeflow setup.
If not, I’m happy to dig deeper. Feel free to reopen!

Cheers,
Max

**Comment by Muhamamd-Rehman** [comment]
Hi @mrokuss 
Thanks a lot for looking into the issue. 

Indeed the solution you gave solves the issue.

But, I still need to keep `-npp` and `-nps` to 1. Increasing them would again cause the same issue.

Also, I encountered the similar RAM issue while concatenating the `data` and `dense_prompt` tensors. Both of them are of size `(1, 980, 625, 625)`

`data = torch.cat([data, dense_prompt], dim=0)`

For time being, I moved both the tensors to GPU before making the inference. Nonetheless, I received the **OOM lack of memory error** while making prediction for the first lesion.

If possible, Could you share the approximate **RAM** requirements and **GPU specifications** (number of GPUs + memory per GPU) needed to run the script?

I will then try to request our IT team to increase my notebook resources.

Once again thanks a a lot for your consideration.

**Comment by mrokuss** [comment]
Hi @Muhamamd-Rehman 

glad to hear the solution helped 👍

Regarding your question about exact RAM / GPU requirements: unfortunately I can’t give reliable numbers. Memory usage depends very strongly on the image sizes, so there isn’t a single set of requirements that fits all cases.

That said, I can share my setup for reference. I’ve been running the script without issues on:

- 64 GB system RAM (this can be increased artificially by adding more swap)
- NVIDIA GeForce RTX 3090 with 24 GB VRAM

I’d suggest asking your IT team for as much RAM and GPU memory as reasonably possible, especially if you plan to work with large images or increase -npp / -nps.

Hope this helps, and feel free to follow up if you have more questions.

**Comment by Muhamamd-Rehman** [comment]
Hi @mrokuss Thanks for sharing your resources.

My notebook resources has been increased to 64 GB RAM and 8 CPU cores. Now the model is running smoothly with all default settings. 
I really appreciate your help and guidance!

Regards,
Muhammad
